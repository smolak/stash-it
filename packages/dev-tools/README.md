# `@stash-it/dev-tools`

![logo-stash-it-color-dark 2x](https://user-images.githubusercontent.com/1819138/30385483-99fd209c-98a7-11e7-85e2-595791d8d894.png)

Set of tools for development purposes.

## Table of Contents

- [Installation](#installation)
- [Usage](#usage)
- [License](#license)
- [Contribution](#contribution)

npm
```bash
npm install @stash-it/dev-tools
```

deno
```bash
deno add @stash-it/dev-tools
```

yarn
```bash
yarn dlx jsr add @stash-it/dev-tools
```

pnpm
```bash
pnpm dlx jsr add @stash-it/dev-tools
```

bun
```bash
bunx jsr add @stash-it/dev-tools
```

## Usage

### `runAdapterTests`

If you want to create an adapter, it should pass all of those tests. Here's how you can use them.
Let's say you use vitest:

```typescript
// Vitest stuff
import { describe } from "vitest";

import { adapterTestCases } from '@stash-it/dev-tools';

// Your adapter.
import { YourAdapter } from './your-adapter';

describe('YourAdapter', () => {
  const adapter = new YourAdapter();
  
  // Add whatever validation or other stuff you need.
  
  // Run the tests.
  adapterTestCases(adapter);
});
```

This helper function accepts an adapter instance and runs tests on it. It will throw an error if any of the tests fail.
It is also capable of cleaning up after self. Meaning, if you're operating on a persistent storage, it will delete all of the
items created during the tests.

### `getHandler`

If you want to creat a plugin, you can use this helper function to get a handler for a specific hook.
Here's how you can use it:

```typescript
// Vitest stuff
import { describe, it, expect } from "vitest";

import { getHandler } from '@stash-it/dev-tools';

// Your plugin.
import { createPlugin } from './your-plugin';

describe('YourPlugin', () => {
  it('does something on buildKey', async () => {
    const plugin = createPlugin();
    const key = "key";
    const handler = getHandler('buildKey', plugin);
    
    // Here is where you can test the handler that you created.
    const result = await handler(key);
    
    // expect ... etc.
  });
});
```

Why not simply do this?

```typescript
const handler = plugin.hookHandlers.buildKey;
```

Well ... current types set around plugins are not that great.
This helper function will throw an error if the handler is not a function. Also, thanks to the argument passed (hook name),
it will know what arguments the handler should accept etc. And you don't need to reach to plugin internals. You simply
request a handler for a specific hook and that's it.

## License

MIT

## Contribution

Feel free to open an issue or a pull request.
