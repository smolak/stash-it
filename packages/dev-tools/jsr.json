{
  "name": "@stash-it/dev-tools",
  "version": "0.0.0-beta.4",
  "exports": "./mod.ts",
  "publish": {
    "include": ["package.json", "README.md", "mod.ts", "src/**/*.ts"],
    "exclude": ["src/**/*.test.ts"]
  }
}
