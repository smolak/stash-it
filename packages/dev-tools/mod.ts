export * from "./src/test/adapter-test-cases";
export * from "./src/test/helpers";
