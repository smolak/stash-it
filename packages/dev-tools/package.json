{
  "name": "@stash-it/dev-tools",
  "version": "0.0.5",
  "type": "module",
  "publishConfig": {
    "access": "public"
  },
  "exports": "./mod.ts",
  "files": [
    "mod.ts",
    "src",
    "_dist",
    "README.md",
    "package.json"
  ],
  "license": "MIT",
  "keywords": [
    "@stash-it",
    "stash it",
    "stash-it",
    "dev-tools",
    "tools"
  ],
  "scripts": {
    "lint": "eslint .",
    "publish": "pnpx jsr publish"
  },
  "dependencies": {
    "@stash-it/core": "npm:@jsr/stash-it__core@0.3.0",
    "nanoid": "^5.0.8",
    "vitest": "^2.1.4"
  }
}
