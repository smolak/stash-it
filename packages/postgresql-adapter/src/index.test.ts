import { describe, expect, it, afterAll } from "vitest";
import { runAdapterTests } from "@stash-it/dev-tools";
import { Client } from "pg";

import { PostgreSqlAdapter, postgreSqlAdapterConfigurationSchema, type PostgreSqlAdapterConfiguration } from "./index";

const connectionConfiguration: PostgreSqlAdapterConfiguration["connection"] = {
  host: "localhost",
  user: "user",
  password: "password",
  database: "dbname",
  port: 5432,
};

const tableName = "items";
const keyColumnName = "key";
const valueColumnName = "value";
const extraColumnName = "extra";

const tableConfiguration: PostgreSqlAdapterConfiguration["table"] = {
  tableName,
  keyColumnName,
  valueColumnName,
  extraColumnName,
};

const adapterConfiguration: PostgreSqlAdapterConfiguration = {
  connection: connectionConfiguration,
  table: tableConfiguration,
};

const prepareDatabase = async (configuration: PostgreSqlAdapterConfiguration) => {
  const client = new Client(configuration.connection);

  await client.connect();

  await client.query(`DROP TABLE IF EXISTS "${tableName}"`);

  await client.query(
    `CREATE TABLE "${tableName}" (
      "${keyColumnName}" TEXT PRIMARY KEY,
      "${valueColumnName}" JSONB NOT NULL, 
      "${extraColumnName}" JSONB NOT NULL
     )`,
  );

  await client.end();
};

describe("PostgreSqlAdapter", async () => {
  afterAll(async () => {
    const client = new Client(adapterConfiguration.connection);

    await client.connect();
    await client.query(`DROP TABLE IF EXISTS "${tableName}"`);
    await client.end();
  });

  describe("configuration", () => {
    describe("connection", () => {
      describe("host", () => {
        it("throws an error when host is not provided", () => {
          expect(
            () =>
              new PostgreSqlAdapter({
                connection: { ...connectionConfiguration, host: undefined as unknown as string },
              }),
          ).toThrowErrorMatchingSnapshot();
        });

        it("throws an error when host is empty", () => {
          expect(
            () => new PostgreSqlAdapter({ connection: { ...connectionConfiguration, host: "" } }),
          ).toThrowErrorMatchingSnapshot();
        });
      });

      describe("user", () => {
        it("throws an error when user is not provided", () => {
          expect(
            () =>
              new PostgreSqlAdapter({
                connection: { ...connectionConfiguration, user: undefined as unknown as string },
              }),
          ).toThrowErrorMatchingSnapshot();
        });

        it("throws an error when user is empty", () => {
          expect(
            () => new PostgreSqlAdapter({ connection: { ...connectionConfiguration, user: "" } }),
          ).toThrowErrorMatchingSnapshot();
        });
      });

      describe("password", () => {
        it("throws an error when password is not provided", () => {
          expect(
            () =>
              new PostgreSqlAdapter({
                connection: { ...connectionConfiguration, password: undefined as unknown as string },
              }),
          ).toThrowErrorMatchingSnapshot();
        });

        it("throws an error when password is empty", () => {
          expect(
            () => new PostgreSqlAdapter({ connection: { ...connectionConfiguration, password: "" } }),
          ).toThrowErrorMatchingSnapshot();
        });
      });

      describe("database", () => {
        it("throws an error when database is not provided", () => {
          expect(
            () =>
              new PostgreSqlAdapter({
                connection: { ...connectionConfiguration, database: undefined as unknown as string },
              }),
          ).toThrowErrorMatchingSnapshot();
        });

        it("throws an error when database is empty", () => {
          expect(
            () => new PostgreSqlAdapter({ connection: { ...connectionConfiguration, database: "" } }),
          ).toThrowErrorMatchingSnapshot();
        });
      });

      describe("port number", () => {
        it("should default to 5432 when port is not provided", () => {
          const {
            connection: { port },
          } = postgreSqlAdapterConfigurationSchema.parse({
            connection: { ...connectionConfiguration, port: undefined },
          });

          expect(port).toBe(5432);
        });

        it("throws an error when port is not a number", () => {
          expect(
            () =>
              new PostgreSqlAdapter({
                connection: { ...connectionConfiguration, port: "not_a_port_number" as unknown as number },
              }),
          ).toThrowErrorMatchingSnapshot();
        });
      });
    });

    describe("table", () => {
      it("should use default table configuration when it is not provided", () => {
        const { table } = postgreSqlAdapterConfigurationSchema.parse({
          ...adapterConfiguration,
          table: undefined,
        });

        expect(table).toEqual({
          tableName: "items",
          keyColumnName: "key",
          valueColumnName: "value",
          extraColumnName: "extra",
        });
      });

      describe("tableName", () => {
        it('should default to "items" if not provided', () => {
          const {
            table: { tableName },
          } = postgreSqlAdapterConfigurationSchema.parse({
            connection: connectionConfiguration,
            table: { tableName: undefined },
          });

          expect(tableName).toEqual("items");
        });

        it("should throw if empty", () => {
          expect(
            () => new PostgreSqlAdapter({ ...adapterConfiguration, table: { tableName: "" } }),
          ).toThrowErrorMatchingSnapshot();
        });

        it("should throw if not a string", () => {
          expect(
            () => new PostgreSqlAdapter({ ...adapterConfiguration, table: { tableName: 1 as unknown as string } }),
          ).toThrowErrorMatchingSnapshot();
        });
      });

      describe("keyColumnName", () => {
        it('should default to "key" if not provided', () => {
          const {
            table: { keyColumnName },
          } = postgreSqlAdapterConfigurationSchema.parse({
            connection: connectionConfiguration,
            table: { keyColumnName: undefined },
          });

          expect(keyColumnName).toEqual("key");
        });

        it("should throw if empty", () => {
          expect(
            () => new PostgreSqlAdapter({ ...adapterConfiguration, table: { keyColumnName: "" } }),
          ).toThrowErrorMatchingSnapshot();
        });

        it("should throw if not a string", () => {
          expect(
            () => new PostgreSqlAdapter({ ...adapterConfiguration, table: { keyColumnName: 1 as unknown as string } }),
          ).toThrowErrorMatchingSnapshot();
        });
      });

      describe("valueColumnName", () => {
        it('should default to "value" if not provided', () => {
          const {
            table: { valueColumnName },
          } = postgreSqlAdapterConfigurationSchema.parse({
            connection: connectionConfiguration,
            table: { valueColumnName: undefined },
          });

          expect(valueColumnName).toEqual("value");
        });

        it("should throw if empty", () => {
          expect(
            () => new PostgreSqlAdapter({ ...adapterConfiguration, table: { valueColumnName: "" } }),
          ).toThrowErrorMatchingSnapshot();
        });

        it("should throw if not a string", () => {
          expect(
            () =>
              new PostgreSqlAdapter({ ...adapterConfiguration, table: { valueColumnName: 1 as unknown as string } }),
          ).toThrowErrorMatchingSnapshot();
        });
      });

      describe("extraColumnName", () => {
        it('should default to "extra" if not provided', () => {
          const {
            table: { extraColumnName },
          } = postgreSqlAdapterConfigurationSchema.parse({
            connection: connectionConfiguration,
            table: { extraColumnName: undefined },
          });

          expect(extraColumnName).toEqual("extra");
        });

        it("should throw if empty", () => {
          expect(
            () => new PostgreSqlAdapter({ ...adapterConfiguration, table: { extraColumnName: "" } }),
          ).toThrowErrorMatchingSnapshot();
        });

        it("should throw if not a string", () => {
          expect(
            () =>
              new PostgreSqlAdapter({ ...adapterConfiguration, table: { extraColumnName: 1 as unknown as string } }),
          ).toThrowErrorMatchingSnapshot();
        });
      });
    });

    describe("table check", () => {
      it("should throw when table does not exist", async () => {
        const configuration: PostgreSqlAdapterConfiguration = {
          ...adapterConfiguration,
          table: {
            ...adapterConfiguration.table,
            tableName: "non_existent_table",
          },
        };

        const adapter = new PostgreSqlAdapter(configuration);

        await expect(adapter.checkStorage()).rejects.toThrow('relation "non_existent_table" does not exist');
      });

      it("should throw when key column doesn't exist", async () => {
        const configuration: PostgreSqlAdapterConfiguration = {
          ...adapterConfiguration,
          table: {
            ...adapterConfiguration.table,
            keyColumnName: "non_existent_key_column",
          },
        };

        const adapter = new PostgreSqlAdapter(configuration);

        await expect(adapter.checkStorage()).rejects.toThrow('column "non_existent_key_column" does not exist');
      });

      it("should throw when value column doesn't exist", async () => {
        const configuration: PostgreSqlAdapterConfiguration = {
          ...adapterConfiguration,
          table: {
            ...adapterConfiguration.table,
            valueColumnName: "non_existent_value_column",
          },
        };

        const adapter = new PostgreSqlAdapter(configuration);

        await expect(adapter.checkStorage()).rejects.toThrow(
          'column "non_existent_value_column" of relation "items" does not exist',
        );
      });

      it("should throw when extra column doesn't exist", async () => {
        const configuration: PostgreSqlAdapterConfiguration = {
          ...adapterConfiguration,
          table: {
            ...adapterConfiguration.table,
            extraColumnName: "non_existent_extra_column",
          },
        };

        const adapter = new PostgreSqlAdapter(configuration);

        await expect(adapter.checkStorage()).rejects.toThrow(
          'non_existent_extra_column" of relation "items" does not exist',
        );
      });
    });
  });

  describe("adapter tests", async () => {
    await prepareDatabase(adapterConfiguration);

    const adapter = new PostgreSqlAdapter(adapterConfiguration);

    runAdapterTests(adapter);
  });
});
