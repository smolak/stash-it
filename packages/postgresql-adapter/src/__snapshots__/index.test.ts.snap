// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`PostgreSqlAdapter > configuration > connection > database > throws an error when database is empty 1`] = `
[ZodError: [
  {
    "code": "too_small",
    "minimum": 1,
    "type": "string",
    "inclusive": true,
    "exact": false,
    "message": "String must contain at least 1 character(s)",
    "path": [
      "connection",
      "database"
    ]
  }
]]
`;

exports[`PostgreSqlAdapter > configuration > connection > database > throws an error when database is not provided 1`] = `
[ZodError: [
  {
    "code": "invalid_type",
    "expected": "string",
    "received": "undefined",
    "path": [
      "connection",
      "database"
    ],
    "message": "Required"
  }
]]
`;

exports[`PostgreSqlAdapter > configuration > connection > host > throws an error when host is empty 1`] = `
[ZodError: [
  {
    "code": "too_small",
    "minimum": 1,
    "type": "string",
    "inclusive": true,
    "exact": false,
    "message": "String must contain at least 1 character(s)",
    "path": [
      "connection",
      "host"
    ]
  }
]]
`;

exports[`PostgreSqlAdapter > configuration > connection > host > throws an error when host is not provided 1`] = `
[ZodError: [
  {
    "code": "invalid_type",
    "expected": "string",
    "received": "undefined",
    "path": [
      "connection",
      "host"
    ],
    "message": "Required"
  }
]]
`;

exports[`PostgreSqlAdapter > configuration > connection > password > throws an error when password is empty 1`] = `
[ZodError: [
  {
    "code": "too_small",
    "minimum": 1,
    "type": "string",
    "inclusive": true,
    "exact": false,
    "message": "String must contain at least 1 character(s)",
    "path": [
      "connection",
      "password"
    ]
  }
]]
`;

exports[`PostgreSqlAdapter > configuration > connection > password > throws an error when password is not provided 1`] = `
[ZodError: [
  {
    "code": "invalid_type",
    "expected": "string",
    "received": "undefined",
    "path": [
      "connection",
      "password"
    ],
    "message": "Required"
  }
]]
`;

exports[`PostgreSqlAdapter > configuration > connection > port number > throws an error when port is not a number 1`] = `
[ZodError: [
  {
    "code": "invalid_type",
    "expected": "number",
    "received": "string",
    "path": [
      "connection",
      "port"
    ],
    "message": "Expected number, received string"
  }
]]
`;

exports[`PostgreSqlAdapter > configuration > connection > user > throws an error when user is empty 1`] = `
[ZodError: [
  {
    "code": "too_small",
    "minimum": 1,
    "type": "string",
    "inclusive": true,
    "exact": false,
    "message": "String must contain at least 1 character(s)",
    "path": [
      "connection",
      "user"
    ]
  }
]]
`;

exports[`PostgreSqlAdapter > configuration > connection > user > throws an error when user is not provided 1`] = `
[ZodError: [
  {
    "code": "invalid_type",
    "expected": "string",
    "received": "undefined",
    "path": [
      "connection",
      "user"
    ],
    "message": "Required"
  }
]]
`;

exports[`PostgreSqlAdapter > configuration > table > extraColumnName > should throw if empty 1`] = `
[ZodError: [
  {
    "code": "too_small",
    "minimum": 1,
    "type": "string",
    "inclusive": true,
    "exact": false,
    "message": "String must contain at least 1 character(s)",
    "path": [
      "table",
      "extraColumnName"
    ]
  }
]]
`;

exports[`PostgreSqlAdapter > configuration > table > extraColumnName > should throw if not a string 1`] = `
[ZodError: [
  {
    "code": "invalid_type",
    "expected": "string",
    "received": "number",
    "path": [
      "table",
      "extraColumnName"
    ],
    "message": "Expected string, received number"
  }
]]
`;

exports[`PostgreSqlAdapter > configuration > table > keyColumnName > should throw if empty 1`] = `
[ZodError: [
  {
    "code": "too_small",
    "minimum": 1,
    "type": "string",
    "inclusive": true,
    "exact": false,
    "message": "String must contain at least 1 character(s)",
    "path": [
      "table",
      "keyColumnName"
    ]
  }
]]
`;

exports[`PostgreSqlAdapter > configuration > table > keyColumnName > should throw if not a string 1`] = `
[ZodError: [
  {
    "code": "invalid_type",
    "expected": "string",
    "received": "number",
    "path": [
      "table",
      "keyColumnName"
    ],
    "message": "Expected string, received number"
  }
]]
`;

exports[`PostgreSqlAdapter > configuration > table > tableName > should throw if empty 1`] = `
[ZodError: [
  {
    "code": "too_small",
    "minimum": 1,
    "type": "string",
    "inclusive": true,
    "exact": false,
    "message": "String must contain at least 1 character(s)",
    "path": [
      "table",
      "tableName"
    ]
  }
]]
`;

exports[`PostgreSqlAdapter > configuration > table > tableName > should throw if not a string 1`] = `
[ZodError: [
  {
    "code": "invalid_type",
    "expected": "string",
    "received": "number",
    "path": [
      "table",
      "tableName"
    ],
    "message": "Expected string, received number"
  }
]]
`;

exports[`PostgreSqlAdapter > configuration > table > valueColumnName > should throw if empty 1`] = `
[ZodError: [
  {
    "code": "too_small",
    "minimum": 1,
    "type": "string",
    "inclusive": true,
    "exact": false,
    "message": "String must contain at least 1 character(s)",
    "path": [
      "table",
      "valueColumnName"
    ]
  }
]]
`;

exports[`PostgreSqlAdapter > configuration > table > valueColumnName > should throw if not a string 1`] = `
[ZodError: [
  {
    "code": "invalid_type",
    "expected": "string",
    "received": "number",
    "path": [
      "table",
      "valueColumnName"
    ],
    "message": "Expected string, received number"
  }
]]
`;
